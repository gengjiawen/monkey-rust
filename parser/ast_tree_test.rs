#[cfg(test)]
mod tests {
    use crate::parse;
    use insta::*;

    pub fn test_ast_tree(name: &str, input: &str) {
        let let_ast = match parse(input) {
            Ok(node) => {
                serde_json::to_string_pretty(&node).unwrap()
            }
            Err(e) => format!("parse error: {}", e[0])
        };
        println!("{}", let_ast);
        assert_snapshot!(name, let_ast);
    }

    #[test]
    fn test_let() {
        let input = "let a = 3";
        test_ast_tree("test_let", input)
    }

    #[test]
    fn test_string() {
        let input = r#""jw""#;
        test_ast_tree("test_string", input)
    }

    #[test]
    fn test_array() {
        let input = "[1, true]";
        test_ast_tree("test_array", input)
    }

    #[test]
    fn test_hash() {
        let input = r#"{"a": 1}"#;
        test_ast_tree("test_hash", input)
    }
}
