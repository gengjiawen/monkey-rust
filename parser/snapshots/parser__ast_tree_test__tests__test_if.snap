---
source: parser/ast_tree_test.rs
expression: "if (x < y) { x } else { y }"
---
{
  "Program": {
    "type": "Program",
    "body": [
      {
        "stmt_type": "Expr",
        "expr_type": "IF",
        "condition": {
          "expr_type": "INFIX",
          "op": {
            "span": {
              "start": 6,
              "end": 7
            },
            "kind": {
              "type": "LT"
            }
          },
          "left": {
            "expr_type": "IDENTIFIER",
            "name": "x",
            "span": {
              "start": 4,
              "end": 5
            }
          },
          "right": {
            "expr_type": "IDENTIFIER",
            "name": "y",
            "span": {
              "start": 8,
              "end": 9
            }
          },
          "span": {
            "start": 4,
            "end": 9
          }
        },
        "consequent": [
          {
            "stmt_type": "Expr",
            "expr_type": "IDENTIFIER",
            "name": "x",
            "span": {
              "start": 13,
              "end": 14
            }
          }
        ],
        "alternate": [
          {
            "stmt_type": "Expr",
            "expr_type": "IDENTIFIER",
            "name": "y",
            "span": {
              "start": 24,
              "end": 25
            }
          }
        ],
        "span": {
          "start": 0,
          "end": 27
        }
      }
    ],
    "span": {
      "start": 0,
      "end": 28
    }
  }
}
